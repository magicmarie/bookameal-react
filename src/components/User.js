import React, { Component } from "react";
import Loader from "react-loader";
import { notify } from "react-notify-toast";
import { Link } from "react-router-dom";
import axiosInstance from "./common/Apicalls";
import WeekDays from "./WeekDays";

// menu card: customer side
const Menu = props => (
  <div className="menu-item">
    <div className="row">
      <div className="col-md-6">{props.mealName}</div>
      <div className="price col-md-2">{props.price}</div>
      <Link
        to="#"
        className="add col-md-4 text-center"
        title="make an order"
        onClick={() => props.handleAddOrder(props.id, props.mealId)}
      >
        +
      </Link>
    </div>
  </div>
);

class UserDashboard extends Component {
  state = {
    menus: [],
    loaded: false,
    today: ""
  };
  //get menu: customer side
  getMenus = menu_day => {
    axiosInstance
      .get(`/user-menus/${menu_day}`)
      .then(response => {
        this.setState({
          menus: response.data.menus,
          loaded: true,
          today: menu_day.charAt(0).toUpperCase() + menu_day.slice(1)
        });
      })
      .catch(error => {
        if (error.response) {
          const { status } = error.response;
          if (status === 404) {
            this.setState({
              menus: []
            });
          } else if (status === 401) {
            localStorage.removeItem("token");
          }
        } else if (error.request) {
          notify.show("Wrong request", "warning", 2500);
        }
      });
  };

  //post order from menu: user
  handleAddOrder = (id, meal_id) => {
    axiosInstance
      .post(`/orders/${id}/${meal_id}`)
      .then(response => {
        console.log(response.data);
        notify.show(response.data.message, "success", 2500);
      })
      .catch(error => {
        if (error.response) {
          const { status } = error.response;
          if (status === 404) {
            this.setState({
              orders: []
            });
          }
        } else if (error.request) {
          notify.show("Wrong request", "error", 2500);
        }
      });
  };
  componentDidMount() {
    const today = this.getCurrentDay();
    this.getMenus(today);
    this.setState({ today });
  }

  getCurrentDay = () => {
    const numberDay = new Date().getDay();
    const days = [
      { key: 1, name: "Monday" },
      { key: 2, name: "Tuesday" },
      { key: 3, name: "Wednesday" },
      { key: 4, name: "Thursday" },
      { key: 5, name: "Friday" },
      { key: 6, name: "Saturday" },
      { key: 7, name: "Sunday" }
    ];
    const today = days.find(day => day.key === numberDay);
    return today.name;
  };

  getMenu = day => {
    day = day.charAt(0).toUpperCase() + day.slice(1);
    const { menus } = this.state;
    this.setState({ currentDay: day, menus });
  };
  render() {
    const { loaded, menus, today } = this.state;
    const menuItems =
      menus.length === 0 ? (
        <div>No Menu Found</div>
      ) : (
        menus.map((menu, index) => (
          <div key={index}>
            <div className="row">
              <h5>
                <strong>{menu.userName}</strong>
              </h5>
            </div>
            {menu.meals.map(meal => (
              <Menu
                id={meal.menu_id}
                mealId={meal.id}
                mealName={meal.name}
                price={meal.price}
                key={meal.id}
                handleAddOrder={this.handleAddOrder}
              />
            ))}
          </div>
        ))
      );

    return (
      <div className="row">
        <div className="col-md-2">
          <div className="header text-center">Days</div>

          <ul className="list-group">
            <WeekDays getMenu={this.getMenus} />
          </ul>
        </div>
        <div className="vl" />
        <div className="col-md-9">
          <h4 className="header text-center">
            {today}
            's Menus
          </h4>
          <Loader loaded={loaded}>
            <div className="h4">{menus.userName}</div>
            <div>{menuItems}</div>
          </Loader>
        </div>
      </div>
    );
  }
}
export default UserDashboard;
