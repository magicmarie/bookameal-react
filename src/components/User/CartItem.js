import React from "react";

// cart item card: customer side
const CartItem = props => (
  <div>
    <tr>
      <th scope="row">{props.mealName}</th>
      <td>{props.price}</td>
      <td>3</td>
      <td>{props.adminName}</td>
      <td>30000UGX</td>
      <td>
        <button
          className="fa fa-check-circle"
          onClick={() => props.handleAddOrder(props.id, props.mealId)}
        />
      </td>
      <td>
        <button className="fa fa-remove" />
      </td>
    </tr>
  </div>
);
export default CartItem;
