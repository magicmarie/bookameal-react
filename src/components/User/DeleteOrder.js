import React from "react";

// delete order modal

const DeleteOrder = props => (
  <div>
    <div id={`deleteOrderModal${props.id}`} className="modal fade">
      <div className="modal-dialog modal-confirm">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-title">Are you sure?</h4>
            <button
              type="button"
              className="close"
              data-dismiss="modal"
              aria-hidden="true"
            >
              &times;
            </button>
          </div>
          <div className="modal-body">
            <p>
              Do you really want to delete <strong>{props.meal_name}</strong>{" "}
              from your orders? This process cannot be undone.
            </p>
          </div>
          <div className="modal-footer">
            <button
              type="button"
              className="btn btn-info"
              data-dismiss="modal"
              id={`cancelModal${props.id}`}
            >
              Cancel
            </button>
            <button
              type="button"
              className="btn btn-danger"
              onClick={() => props.ConfirmDeleteOrder(props.id)}
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
);

export default DeleteOrder;
