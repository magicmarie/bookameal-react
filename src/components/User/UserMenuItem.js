import React from "react";

// menu card: customer side
const UserMenuItem = props => (
  <div className="menu-item">
    <div className="row menuitem">
      <div className="col-sm-6">
        <p>{props.mealName}</p>
      </div>
      <div className="col-sm-3">
        <p>
          <span className="badge badge-secondary">{props.price}</span>
        </p>
      </div>
      <div className="col-sm-3">
        <p>
          <button
            to="#"
            className="btn add"
            title="add to cart"
            onClick={() =>
              props.handleAddCart(
                props.mealId,
                props.mealName,
                props.price,
                props.adminName,
                props.id
              )
            }
          >
            Add to Cart
          </button>
        </p>
      </div>
    </div>
  </div>
);
export default UserMenuItem;
