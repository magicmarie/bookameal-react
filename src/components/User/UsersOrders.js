import React, { Component } from "react";
import Loader from "react-loader";
import { notify } from "react-notify-toast";
import axiosInstance from "../common/Apicalls";
import Pagination from "../common/Pagination";

// order: customer side
const Order = props => (
  <div className="card mb-2 meal">
    <div className="card-body">
      <div className="card-text">
        <strong>Caterer: </strong>
        {props.adminName}
      </div>
      <div className="card-text">
        <strong>Meal name: </strong>
        {props.mealName}
      </div>
      <div className="card-text">
        <strong>Price: </strong>
        {props.price}
        UGX
      </div>
      <div className="card-text">
        <strong>Quantity: </strong>
        {props.quantity}
      </div>
    </div>
  </div>
);

/**
 * @class UserOrders
 * @extends {Component}
 */
class UserOrders extends Component {
  state = {
    orders: [],
    loaded: false,
    nextPage: null,
    previousPage: null,
    currentPage: null,
    pages: null
  };

  /**
   *@returns {orders} orders
   * @memberof UserOrders
   */
  componentDidMount() {
    this.getUserOrders();
  }

  // get user orders
  getUserOrders = () => {
    const page = this.state.currentPage || 1;

    axiosInstance
      .get("/user/orders", { params: { page } })
      .then(response => {
        const {
          currentPage,
          orders,
          nextPage,
          pages,
          previousPage
        } = response.data.Orders;
        this.setState({
          currentPage,
          orders,
          nextPage,
          pages,
          previousPage,
          loaded: true
        });
      })
      .catch(error => {
        if (error.response) {
          const { status } = error.response;
          if (status === 401) {
            localStorage.removeItem("token");
          }
        } else if (error.request) {
          notify.show("Wrong request", "warning", 2500);
        }
      });
  };

  // pagination
  changePage = selectedPage => {
    this.setState(
      {
        currentPage: selectedPage
      },
      () => this.getUserOrders()
    );
  };

  /**
   *
   *
   * @returns {any} rendered items
   * @memberof UserOrders
   */
  render() {
    const {
      orders,
      loaded,
      currentPage,
      pages,
      nextPage,
      previousPage
    } = this.state;
    const ordersDetails =
      orders.length === 0 ? (
        <div>No orders Found</div>
      ) : (
        orders.map(order => (
          <div className="col-md-3 col-lg-3" key={order.id}>
            <Order
              id={order.id}
              mealName={order.meal_name}
              price={order.price}
              quantity={order.quantity}
              adminName={order.adminName}
              ConfirmDeleteOrder={this.ConfirmDeleteOrder}
            />
          </div>
        ))
      );

    return (
      <div>
        <Loader loaded={loaded}>
          <h2 className="header text-center">Orders</h2>
          <div className="row">{ordersDetails}</div>
        </Loader>
        <Pagination
          currentPage={currentPage}
          previousPage={previousPage}
          nextPage={nextPage}
          pages={pages}
          changePage={this.changePage}
        />
      </div>
    );
  }
}
export default UserOrders;
