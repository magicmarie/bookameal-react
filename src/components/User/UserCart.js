import React from "react";
import Loader from "react-loader";
import { notify } from "react-notify-toast";
import axiosInstance from "../common/Apicalls";
import CartItem from "./CartItem";
import { contextWrapper } from "../common/Helper";

/**
 * @class UserCart
 * @extends {React.Component}
 */
export class UserCart extends React.Component {
  state = {
    loaded: true
  };

  /**
   *
   *@returns {cartitems} cartitems
   * @memberof UserCart
   */

  onChange = (evt, mealId) => {
    const quantity = evt.target.value;
    this.props.context.setQuantity(mealId, quantity);
  };

  makeOrder = () => {
    const items = this.props.context.getCart();
    axiosInstance.post("/orders", { meals: items }).then(res => {
      // Redirect to  orders
      notify.show("Order made successfully", "success", 2500);
      this.props.context.clearCart();
      this.props.history.push("/usersorders");
    });
  };

  /**
   * @returns {any} rendered items
   * @memberof UserCart
   */
  render() {
    const { loaded } = this.state;
    const cartItems = this.props.context.getCart();
    const orders =
      cartItems.length === 0 ? (
        <tr>
          <div>No orders added to cart yet</div>
        </tr>
      ) : (
        cartItems.map((order, index) => (
          <CartItem
            key={index}
            id={order.meal_id}
            quantity={order.quantity}
            menuId={order.menu_id}
            mealName={order.meal_name}
            price={order.price}
            adminName={order.adminName}
            onChange={this.onChange}
            removeItem={this.props.context.deleteFromCart}
          />
        ))
      );
    return (
      <div>
        <h4 className="header text-center">My Cart </h4>
        <Loader loaded={loaded}>
          <table className="table table-bordered table-hover">
            <thead className="thead-light">
              <tr>
                <th scope="col">Caterer</th>
                <th scope="col">Meal</th>
                <th scope="col">Quantity</th>
                <th>Unit Price(UGX)</th>
                <th scope="col">Sub Total(UGX)</th>
                <th scope="col">Cancel</th>
              </tr>
            </thead>
            <tbody>{orders}</tbody>
          </table>
          <button
            disabled={cartItems.length === 0}
            className="btn btn-primary"
            onClick={this.makeOrder}
          >
            Make orders
          </button>
        </Loader>
      </div>
    );
  }
}

export default contextWrapper(UserCart);
