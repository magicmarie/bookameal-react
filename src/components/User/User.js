import React, { Component } from "react";
import Loader from "react-loader";
import { notify } from "react-notify-toast";
import axiosInstance from "../common/Apicalls";
import UserMenuItem from "./UserMenuItem";
import getCurrentDay from "../common/CurrentDay";
import { HelpDiv, contextWrapper } from "../common/Helper";

/**
 * @class UserDashboard
 * @extends {Component}
 */
export class UserDashboard extends Component {
  state = {
    menus: [],
    loaded: false,
    today: ""
  };

  /**
   *@returns {menus} menus
   * @memberof UserDashboard
   */
  componentDidMount() {
    const today = getCurrentDay();
    this.getMenus(today);
    this.setState({ today });
  }

  // get menu: customer side
  /**
   *@returns {null} null
   *@param {string} menu_day
   * @memberof UserDashboard
   */
  getMenus = menu_day => {
    axiosInstance
      .get(`/user-menus/${menu_day}`)
      .then(response => {
        this.setState({
          menus: response.data.menus,
          loaded: true,
          today: menu_day.charAt(0).toUpperCase() + menu_day.slice(1)
        });
      })
      .catch(error => {
        if (error.response) {
          const { status } = error.response;
          if (status === 401) {
            localStorage.removeItem("token");
          }
        } else if (error.request) {
          notify.show("Wrong request", "warning", 2500);
        }
      });
  };

  /**
   *
   *@param {string} day
   *@returns {null} null
   * @memberof UserDashboard
   */
  getMenu = day => {
    day = day.charAt(0).toUpperCase() + day.slice(1);
    const { menus } = this.state;
    this.setState({ currentDay: day, menus });
  };

  /**
   *@param meal_id
   *@param meal_name
   *@param adminName
   *@param id
   *@returns {null} null
   * @memberof UserDashboard
   */
  handleAddCart = (meal_id, meal_name, price, adminName, id) => {
    const cartItem = {
      menu_id: id,
      meal_id,
      meal_name,
      price,
      adminName
    };
    this.props.context.setCart(cartItem);
    notify.show("Added to your Cart", "success", 2000);
  };

  /**
   *
   *
   * @returns {any} rendered items
   * @memberof UserDashboard
   */
  render() {
    const { loaded, menus, today } = this.state;
    const menuItems =
      menus.length === 0 ? (
        <div>No Menu Found</div>
      ) : (
        menus.map((menu, index) => (
          <div className="menu" key={index}>
            <p className="text-center">
              <strong>
                Caterer {" : "}
                {menu.userName}
              </strong>
            </p>
            {menu.meals.map(meal => (
              <UserMenuItem
                id={meal.menu_id}
                mealId={meal.id}
                mealName={meal.name}
                price={meal.price}
                adminName={menu.userName}
                key={meal.id}
                handleAddCart={this.handleAddCart}
              />
            ))}
          </div>
        ))
      );

    return (
      <div className="row">
        <HelpDiv getMenu={this.getMenu} />
        <div className="col-md-10">
          <h4 className="header text-center">
            {today}
            's Menus
          </h4>
          <Loader loaded={loaded}>
            <div>{menuItems}</div>
          </Loader>
        </div>
      </div>
    );
  }
}
export default contextWrapper(UserDashboard);
