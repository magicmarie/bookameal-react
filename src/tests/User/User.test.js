import React from "react";
import { shallow } from "enzyme";
import MockAdapter from "axios-mock-adapter";
import { UserDashboard } from "../../components/User/User";
import axiosInstance from "../../components/common/Apicalls";
import getCurrentDay from "../../components/common/CurrentDay";

describe("UserDashboard component", () => {
  let data;
  let wrapper;
  let mockFn;
  beforeEach(() => {
    data = [
      {
        id: 1,
        meals: [{ id: 1, menu_id: 1, name: "test", price: 1000 }],
        userName: "martha"
      }
    ];
    mockFn = jest.fn();
    const context = {
      setCart: mockFn
    };
    wrapper = shallow(<UserDashboard context={context} />);
  });

  it("should handle add to cart", () => {
    wrapper.instance().handleAddCart(1, "test", 1000, "test", 2);
    expect(mockFn.mock.calls.length).toEqual(1);
  });
  it("should handle get menus", async () => {
    // mock axios to handle requests made when component mounts
    const today = getCurrentDay();
    const axiosMock = new MockAdapter(axiosInstance);
    axiosMock.onGet(`/user-menus/${today}`).reply(200, {
      menus: data
    });
  });

  it("should catch errors on get menus ", () => {
    const today = getCurrentDay();
    const axiosMock = new MockAdapter(axiosInstance);
    axiosMock.onGet(`/user-menus/${today}`).reply(401, {});
  });

  it("gets menu", () => {
    const today = getCurrentDay();
    wrapper.instance().getMenu(today);
    const { currentDay } = wrapper.instance().state;
    expect(currentDay).toEqual(today);
  });

  it("renders successfully", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
