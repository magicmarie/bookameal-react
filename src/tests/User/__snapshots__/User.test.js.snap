// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDashboard component renders successfully 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserDashboard
    context={
        Object {
            "setCart": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          getMenu={[Function]}
/>,
        <div
          className="col-md-10"
>
          <h4
                    className="header text-center"
          >
                    Monday
                    's Menus
          </h4>
          <Unknown
                    component="div"
                    loaded={false}
                    loadedClassName="loadedContent"
                    parentClassName="loader"
          >
                    <div>
                              <div>
                                        No Menu Found
                              </div>
                    </div>
          </Unknown>
</div>,
      ],
      "className": "row",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "getMenu": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h4
              className="header text-center"
>
              Monday
              's Menus
</h4>,
            <Unknown
              component="div"
              loaded={false}
              loadedClassName="loadedContent"
              parentClassName="loader"
>
              <div>
                            <div>
                                          No Menu Found
                            </div>
              </div>
</Unknown>,
          ],
          "className": "col-md-10",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Monday",
                "'s Menus",
              ],
              "className": "header text-center",
            },
            "ref": null,
            "rendered": Array [
              "Monday",
              "'s Menus",
            ],
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <div>
                <div>
                                No Menu Found
                </div>
</div>,
              "component": "div",
              "loaded": false,
              "loadedClassName": "loadedContent",
              "parentClassName": "loader",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div>
                  No Menu Found
</div>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "No Menu Found",
                },
                "ref": null,
                "rendered": "No Menu Found",
                "type": "div",
              },
              "type": "div",
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            getMenu={[Function]}
/>,
          <div
            className="col-md-10"
>
            <h4
                        className="header text-center"
            >
                        Monday
                        's Menus
            </h4>
            <Unknown
                        component="div"
                        loaded={false}
                        loadedClassName="loadedContent"
                        parentClassName="loader"
            >
                        <div>
                                    <div>
                                                No Menu Found
                                    </div>
                        </div>
            </Unknown>
</div>,
        ],
        "className": "row",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "getMenu": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h4
                className="header text-center"
>
                Monday
                's Menus
</h4>,
              <Unknown
                component="div"
                loaded={false}
                loadedClassName="loadedContent"
                parentClassName="loader"
>
                <div>
                                <div>
                                                No Menu Found
                                </div>
                </div>
</Unknown>,
            ],
            "className": "col-md-10",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Monday",
                  "'s Menus",
                ],
                "className": "header text-center",
              },
              "ref": null,
              "rendered": Array [
                "Monday",
                "'s Menus",
              ],
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <div>
                  <div>
                                    No Menu Found
                  </div>
</div>,
                "component": "div",
                "loaded": false,
                "loadedClassName": "loadedContent",
                "parentClassName": "loader",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div>
                    No Menu Found
</div>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "No Menu Found",
                  },
                  "ref": null,
                  "rendered": "No Menu Found",
                  "type": "div",
                },
                "type": "div",
              },
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
